quill/ecs/src/storage/sparse_set.rs

```diff
@@ -0,0 +1,118 @@
+use std::{iter, mem::MaybeUninit, ptr::NonNull};
+
+use component::ComponentTypeId;
+
+use crate::component::{self, ComponentMeta};
+
+use super::blob_vec::BlobVec;
+
+/// Stores components in a sparse set.
+pub struct SparseSetStorage {
```

Comment:
`FnvHashMap` is insanely fast, maybe it could be used in stead of `SparseSetStorage`?

Might require some architectural changes, since for each entity you would lookup its components by their id.

Reply:
I'm using sparse sets instead of hash tables because we can optimize query iteration by e.g. packing all entities matching a given archetype into the front of the array. See https://skypjack.github.io/2019-03-07-ecs-baf-part-2/

Reply:
Wouldn't using a HashMap defeat part of the purpose of ECS by getting rid of linear memory reads?

Reply:
Sparse set already gets rid of 100% linear memory reads. Also `FnvHashMap` is not just fast, its insanely fast... it even beats bitsets... its crazy.

Reply:
I mean the memory reads with sparse sets are still linear, just with gaps.
Using a HashMap will make memory reads completely unpredictable.

Reply:
Hmm now I'm wondering where I saw `next_power_of_two` being used. `RawVec` uses the following algorithm
```rust
let required_cap = used_cap + needed_extra_cap;
let double_cap = self.cap * 2;
let new_cap = cmp::max(required_cap, double_cap);
```

Reply:
These two code snippets produce different behaviour. Also you are in the wrong comment thread xD

Reply:
Everything just seems to be going wrong for me at the moment 😅

Yea, the behavior is slightly different but both ensures push are `O(n)` amortized. The implementation that does not use `next_power_two` will double the capacity unless the total required cap is greater than the doubled size. Where the `next_power_of_two` will double and if the capacity is greater than `cap * 2` it will be `cap * 2 * 2` it will repeat this until double cap is equal or greater than required cap.